src/core/utils/unique_service_manager.cpp

```diff
@@ -0,0 +1,299 @@
+    mEntries[i].mServerCompareCallback = aServerCompareCallback != NULL ? aServerCompareCallback : DefaultServerCompare;
+    mEntries[i].mContext               = aContext;
+
+    if (NetworkDataLeaderServiceLookup(i, config, rlocIn) == OT_ERROR_NONE)
+    {
+        mEntries[i].mServiceUpdateCallback(&config, mEntries[i].mContext);
+    }
+    else
+    {
+        AddNetworkDataLocalService(*mEntries[i].mServiceConfig);
+        GetNetif().GetNetworkDataLocal().SendServerDataNotification();
+    }
+
+exit:
+    return error;
+}
+
+otError UniqueServiceManager::UnregisterService(uint32_t       aEnterpriseNumber,
+                                                const uint8_t *aServiceData,
+                                                uint8_t        aServiceDataLength)
+{
+    otError         error = OT_ERROR_NONE;
+    otServiceConfig config;
+    bool            rlocIn;
+    uint8_t         i;
+
+    for (i = 0; i < OT_ARRAY_LENGTH(mEntries); i++)
+    {
+        if (mEntries[i].mServiceConfig != NULL && mEntries[i].mServiceConfig->mEnterpriseNumber == aEnterpriseNumber &&
+            mEntries[i].mServiceConfig->mServiceDataLength == aServiceDataLength &&
+            memcmp(mEntries[i].mServiceConfig->mServiceData, aServiceData, aServiceDataLength) == 0)
+        {
+            break;
+        }
+    }
+
+    VerifyOrExit(i < OT_ARRAY_LENGTH(mEntries), error = OT_ERROR_NOT_FOUND);
+
+    if (NetworkDataLeaderServiceLookup(i, config, rlocIn) == OT_ERROR_NONE && rlocIn)
+    {
+        RemoveNetworkDataLocalService(*mEntries[i].mServiceConfig);
+        GetNetif().GetNetworkDataLocal().SendServerDataNotification();
+    }
+
+    memset(&mEntries[i], 0, sizeof(ServiceEntry));
+
+exit:
+    return error;
+}
+
+void UniqueServiceManager::HandleStateChanged(Notifier::Callback &aCallback, otChangedFlags aFlags)
+{
+    aCallback.GetOwner<UniqueServiceManager>().HandleStateChanged(aFlags);
+}
+
+void UniqueServiceManager::HandleStateChanged(otChangedFlags aFlags)
+{
+    otServiceConfig config;
+    bool            rlocIn;
+    bool            startTimer = false;
+
+    VerifyOrExit((aFlags & OT_CHANGED_THREAD_NETDATA) != 0);
+
+    for (uint8_t i = 0; i < OT_ARRAY_LENGTH(mEntries); i++)
+    {
+        if (mEntries[i].mServiceConfig == NULL)
+        {
+            continue;
+        }
+
+        if (NetworkDataLeaderServiceLookup(i, config, rlocIn) == OT_ERROR_NONE)
+        {
+            // If the server is not selected as the primary server, it should unregister the service.
+            if (rlocIn && config.mServerConfig.mRloc16 != GetNetif().GetMle().GetRloc16())
+            {
+                // MLE layer will send SVR_DATA.ntf when it detects the mismatch between NetworkDataLeader and
+                // NetworkDataLocal in the function HandleStateChanged().
+                RemoveNetworkDataLocalService(*mEntries[i].mServiceConfig);
+            }
+
+            mEntries[i].mServiceUpdateCallback(&config, mEntries[i].mContext);
```

Comment:
should we clear `mEntries[i]` here?

Reply:
No. Only when the user call `UnregisterService ()`, the `mEntries[i]` is deleted.